### DISCUSSION TITLE
Resize windows to optimize screen usage

### DISCUSSION THREAD

**Discussion by Hoffmano** [paperwm/PaperWM#965] [Ideas]
I'm using PaperWM, and I've noticed that some of my windows are getting cropped, with part of them occupying the edge of my monitor. It's becoming an issue as I can't fully see or interact with parts of the window. Is there a way to resize or adjust window behavior in PaperWM to prevent this from happening? Any suggestions on how to fix this would be appreciated!

My dream would that the PaperWM optimize the use of my screen space

**Comment by Hoffmano** [comment]
Case 1

How is today:
![image](https://github.com/user-attachments/assets/3600374f-4535-4773-9e36-6f65edfe9cef)

How I would like to be:
![image](https://github.com/user-attachments/assets/4d934fdc-49ba-4cf1-b908-ebbb3f59f27b)

Case 2

How is today:
![image](https://github.com/user-attachments/assets/3c62d642-04c8-4026-a461-433f623932c0)

How I would like to be:
![image](https://github.com/user-attachments/assets/a9405e17-6880-4281-a716-7a2b37893138)

**_My dream would that the PaperWM optimize the use of my screen space_**

**Comment by Lythenas** [comment]
Since PaperWM is a _scrolling_ window manager, this is how it is intended.

If you want to maximize a window, you can do so with `Super`+`f`. Then it will take up the entire screen like you want in case 1.

For case 2 you have to "manually" resize the two windows to 50%. This can be done using `Super`+`r` to cycle through the configured "Useful window widths". These can be configured in the settings.

There is also a setting "Horizontal margin" that you may be bothered by. But it wasn't an issue for be before.

Also we have "Window focus modes" (configurable by the icon in the top bar) that change a bit what happens when you focus a window. The first picture in case 1 looks like you may have activated "Center" there. Personally I just leave it as "Default".

**Reply by Hoffmano** [reply]
@Lythenas
I’d greatly appreciate your input.

I understand that PaperWM is a scrolling window manager, but as a product manager, I don’t fully see the value in having cropped windows on my screen. To me, a scrolling window manager should involve windows that are either horizontally or vertically aligned and can be scrolled, which aligns with my proposed approach.

I don’t believe we need to be limited by non-optimal concepts when a higher value-added approach exists.

That said, as a collaborator, could you provide me with a guide on how to fork PaperWM and implement my vision?

**Reply by Lythenas** [reply]
To be honest, I think maybe PaperWM isn't the right extension for you. In PaperWM the scrolling part is quite fundamental and there are a bunch of hacks to make that work. So not sure if forking it would be worth it. It would probably be possible to implement what you want but it may be a bunch of effort to fix all the edge cases. You would probably want to take a look at the function `insertWindow` in tiling.js as a starting point and find a way to resize all windows accordingly after you insert it.

If you want a "normal" tiling window manager in gnome (something that works more like e.g. i3) there are other extensions that come closer to that. It has been a while since I tried any of them but there is e.g. Material Shell, gTile and forge.

Last time I tried them was on Gnome 42 and I ran into problems with all of them. PaperWM was the one where I had the least problems :D And unfortunately I'm still stuck on Gnome 42 at work. I would expect that the extensions are improved on newer gnome versions, so maybe it is worth for you to try them out.

**Reply by Hoffmano** [reply]
Alright, I get it.

I've tried all the extensions you mentioned, and you're right—Material Shell is exactly what I was looking for. Unfortunately, the project is unmaintained and has serious performance issues.

So, I've decided that PaperWM is a good starting point, as it works really well.

I'll try to customize it according to my ideas using your advice, and I’ll definitely be back here for more help if I need it!

**Reply by Hoffmano** [reply]
@Lythenas 
Modify the code into .local/share/gnome-shell/extensions/paperwm@paperwm.github.com should work?
How do you dev and test?
This is my first try to modify a gnome extension

I've tried to modify some parameters and test, but doesn't work :(

**Reply by Hoffmano** [reply]
I have success on my first mods

**Reply by Lythenas** [reply]
That's awesome.

I think we would probably accept a PR for the behavior as long as you can turn it on/off in the settings.

Just FYI, for the most part I also use PaperWM like you want to (either one fullscreen window or two windows side by side). I usually don't have much use for more smaller windows on one monitor since I have two monitors anyway. Currently I just set up the windows manually using the shortcuts. If you can make that easier I wouldn't be opposed to it :D Also I usually still scroll instead of using multiple workspaces. Because I find it a bit quicker. Quite often I just want to switch one of the two half screen windows. So I have the one I want to keep in the center and then scroll left/right depending on which other window I need at the time (example workspace: MS Teams | Terminal | Browser with docs, editor fullscreen on the other monitor).

**Reply by Hoffmano** [reply]
@Lythenas how to know how many windows are rendered on my screen?

Has any prop to this?

**Reply by Lythenas** [reply]
In PaperWM, the class `Space` is the list of windows (or more correctly it is the list of columns that then contain windows). I say "is" because it extends Array. In PaperWM we have some hacks so one workspace is only on one monitor. Maybe take a look at `Space.layout`: https://github.com/paperwm/PaperWM/blob/ad90e0cf7ac69eceecd8fd511acdd5d0dcc14472/tiling.js#L650 It's possible that this function is a good place to resize your windows. But not really sure.

Because some windows can scroll out of the screen, they may not actually be visible. I think you could check that manually using the `workArea` of the space.

In gnome you can also get to the windows somehow. To get to the gnome workspaces you could e.g. use `global.workspace_manager`. Not sure if that will be directly useful for you. There is some limited documentation here: https://gjs-docs.gnome.org/ But it's possible you may need to dig through the source code for some details: https://gitlab.gnome.org/GNOME/gnome-shell

**Reply by Hoffmano** [reply]
I create a [draft](https://github.com/paperwm/PaperWM/pull/967) so you can give me some feedbacks and test the feature

**Comment by jtaala** [comment]
Thanks @Hoffmano,

So, check out PaperWM's readme - windows can be set (by default)) to implement various behaviours you describe.

Checkout [winprops](https://github.com/paperwm/PaperWM?tab=readme-ov-file#setting-window-specific-properties)'s `preferredWidth` setting.  For example, my one looks like:

![image](https://github.com/user-attachments/assets/b42ed1ff-2bc2-45bd-aa8f-b0010c1a2703)

You can use `*` for a default window rule (e.g. can set to 50% or even 100%).

Next up, you can set a [window focus mode](https://github.com/paperwm/PaperWM?tab=readme-ov-file#window-focus-modes) - see the `EDGE` mode, which will ensure that window are always at the edge (and if you have a `preferredWidth` of 50% then things will looks like the last picture you put up).

It does sound though that what you're after isn't quite what PaperWM is designed for.

Still will be interesting to see how this approach might work.

Cheers.
